llvm/lib/Target/EVM/EVMStackSolver.cpp

```diff
@@ -201,54 +259,184 @@ llvm::calculateStackTransformCost(Stack Source, const Stack &Target,
+              if (I > StackDepthLimit)
+                UnreachableSlots.push_back({From, From.size() - I - 1});
+            },
+            // Push or dup.
+            [&](const StackSlot *Slot) {
+              if (Slot->isRematerializable() || isSpillReg(Slot))
+                return;
+
+              auto SlotIt = llvm::find(llvm::reverse(From), Slot);
+              assert(SlotIt != From.rend() &&
+                     "Slot for duplication not found on the stack.");
+              unsigned Depth = std::distance(From.rbegin(), SlotIt);
+              if (Depth >= StackDepthLimit)
+                UnreachableSlots.push_back({From, From.size() - Depth - 1});
+            },
+            // Pop.
+            [&]() {});
+      };
+
+  for (const auto &MBB : MF) {
+    Stack CurrentStack = StackModel.getMBBEntryStack(&MBB);
+
+    // Check the stack transformation for each instruction.
+    for (const auto &MI : StackModel.instructionsToProcess(&MBB)) {
+      checkStackTransformation(CurrentStack, StackModel.getInstEntryStack(&MI));
+      CurrentStack = getMIExitStack(&MI);
+    }
+
+    auto [BranchTy, TBB, FBB, BrInsts, Condition] = getBranchInfo(&MBB);
+
+    // Check the transformation for the exit stack. This mimics logic from
+    // the EVMStackifyCodeEmitter::run() method, as in some cases this is
+    // not needed.
+    switch (BranchTy) {
+    case EVMInstrInfo::BT_Uncond:
+    case EVMInstrInfo::BT_NoBranch:
+      if (!MBB.succ_empty())
+        // For loop latches, exit stack is not the same as the entry stack
+        // of the successor block, so we need to check the transformation
+        // from the current stack to the entry stack of the successor block.
+        checkStackTransformation(CurrentStack,
+                                 StackModel.getMBBEntryStack(TBB));
+      break;
+    case EVMInstrInfo::BT_None:
+      if (!MBB.isReturnBlock())
+        break;
+      LLVM_FALLTHROUGH;
+    case EVMInstrInfo::BT_Cond:
+    case EVMInstrInfo::BT_CondUncond:
+      checkStackTransformation(CurrentStack, StackModel.getMBBExitStack(&MBB));
+      break;
+    }
+  }
+  return UnreachableSlots;
+}
+
+void EVMStackSolver::run() {
+  unsigned IterCount = 0;
+  while (true) {
+    runPropagation();
+    LLVM_DEBUG({
+      dbgs() << "************* Stack *************\n";
+      dump(dbgs());
+    });
+
+    auto UnreachableSlots = getUnreachableSlots();
+    if (UnreachableSlots.empty())
+      break;
+
+    if (MF.getFunction().hasFnAttribute("evm-recursive")) {
+      // For recursive functions we can't use spills to fix the stack too deep
+      // errors, as we are using memory to spill and not real stack. Report an
+      // error if this function is recursive and we forced compress stack
+      // before.
+      if (ForceCompressStack)
+        report_fatal_error(
+            "Stackification failed for '" + MF.getName() +
+            "' function. It is recursive and has stack too deep errors.");
```

Comment:
Is it worth to add something like "Consider refactoring it to use a non-recursive approach." to the message?